<?xml version="1.0"?>
<!--for usv, get yaw from imu-->
<launch>
  <node name="joy" pkg="joy" type="joy_node" />
  <node name="light" pkg="mz_usv" type="lights_io" />
  <node name="heartbeat_sender" pkg="mz_usv" type="heartbeat_sender" />
  <node name="VI_sensor" pkg="mz_usv" type="VI_sensor" />
  <node name="imu" pkg="mz_usv" type="imu" output="screen" />
  <node name="imu_filter_node" pkg="imu_filter_madgwick" type="imu_filter_node" >
	<param name = "constant_dt"  value ="0.0"/>
	<remap from="imu/data" to="/sensor/m_imu"/>
	<param name = "world_frame"  value ="enu"/>	
	<param name = "use_mag"  value ="true"/>	
	<param name = "use_magnetic_field_msg"  value ="true"/>		
  </node>
  <!--odometry  /-->
  <node name="bl_map" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 1 map odom" />
  <node name="bl_odom" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 1 odom base_link" />
  <node name="bl_gps" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 1 odom gps" />
  <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se_odom" clear_params="true">
    <param name="frequency" value="51"/>
    <param name="sensor_timeout" value="2.0"/>
    <param name="two_d_mode" value="true"/>
    <param name="map_frame" value="map"/>
    <param name="odom_frame" value="odom"/>
    <param name="base_link_frame" value="base_link"/>
    <param name="world_frame" value="odom"/>
    <param name="publish_tf" value="true"/>
    <param name="imu0" value="/sensor/m_imu"/>
    <rosparam param="imu0_config">[false, false, false,
                                   false,  false,  true,
                                   false,  false,  false,
                                   false,  false,  true,
                                   false,  false,  false]</rosparam>
    <param name="imu0_differential" value="false"/>
    <param name="imu0_remove_gravitational_acceleration" value="true"/>

    <param name="odom0" value="/odometry/gps"/>
    <rosparam param="odom0_config">[true,  true,  true,
                                   false, false, false,
                                   false, false, false,
                                   false, false, false,
                                   false, false, false]</rosparam>
	<rosparam param="process_noise_covariance">[1e-3, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                             0,    1e-3, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                             0,    0,    1e-3, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                             0,    0,    0,    1e-8, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                             0,    0,    0,    0,    1e-8, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                             0,    0,    0,    0,    0,    1e-3, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                             0,    0,    0,    0,    0,    0,    1e-8, 0,     0,    0,    0,    0,    0,    0,    0,
                             0,    0,    0,    0,    0,    0,    0,     1e-8, 0,    0,    0,    0,    0,    0,    0,
                             0,    0,    0,    0,    0,    0,    0,     0,     1e-3, 0,    0,    0,    0,    0,    0, 
                             0,    0,    0,    0,    0,    0,    0,     0,     0,    1e-8, 0,    0,    0,    0,    0,
                             0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    1e-8, 0,    0,    0,    0,
                             0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    1e-8, 0,    0,    0,
                             0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    1e-8, 0,    0,
                             0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    1e-8, 0,
                             0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    1e-8]</rosparam>
		<rosparam param="initial_estimate_covariance">[1e-2, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                0,    1e-2, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                0,    0,    1e-2, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                0,    0,    0,    1e-7, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                0,    0,    0,    0,    1e-7, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                0,    0,    0,    0,    0,    1e-2, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                0,    0,    0,    0,    0,    0,    1e-7, 0,    0,    0,     0,     0,     0,    0,    0,
                                0,    0,    0,    0,    0,    0,    0,    1e-7, 0,    0,     0,     0,     0,    0,    0,
                                0,    0,    0,    0,    0,    0,    0,    0,    1e-2, 0,     0,     0,     0,    0,    0,
                                0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-7,  0,     0,     0,    0,    0,
                                0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-7,  0,     0,    0,    0,
                                0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-7,  0,    0,    0,
                                0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-7, 0,    0,
                                0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-7, 0,
                                0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-7]</rosparam>
    <param name="odom0_differential" value="false"/>
 </node>
  <node pkg="robot_localization" type="navsat_transform_node" name="navsat_transform" clear_params="true">
    <param name="frequency" value="1"/>
    <param name="delay" value="0.2"/>
    <param name="magnetic_declination_radians" value="0"/>
    <param name="yaw_offset" value="0"/>
    <param name="zero_altitude" value="true"/>
    <param name="broadcast_utm_transform" value="true"/>
    <param name="wait_for_datum" value="false"/>
    <remap from="imu/data" to="/sensor/m_imu"/>
      <remap from="gps/fix" to="/sensor/m_fix"/>
     <remap from="odometry/filtered" to="odometry/filtered"/>
  </node>
  <!--usv  /-->
  <node name="auto_pwm" pkg="mz_usv" type="auto_pwm" launch-prefix="sudo -E" output="screen" /> 
  <!--lidar-->
  <node name="bl_lidar" pkg="tf2_ros" type="static_transform_publisher" args="0 0 0 0 0 0 1  base_link laser" />
  <!--<node name="rplidarNode"          pkg="rplidar_ros"  type="rplidarNode" output="screen">
    <param name="serial_port"         type="string" value="/dev/ttyUSB0"/>  
    <param name="serial_baudrate"     type="int"    value="1000000"/>
    <param name="frame_id"            type="string" value="laser"/>
    <param name="inverted"            type="bool"   value="false"/>
    <param name="angle_compensate"    type="bool"   value="true"/>
  </node>-->	
</launch>


